### ISSUE TITLE
Refactor remaining examples into architectural blocks with thoughtful DRY boundaries

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#119]
Follow-up to #116 after the `define-layer` syntax merge. Refactor the remaining examples into meaningful architectural components so the top-level model reads like the diagram and repeated architectural behavior has a clear home.

## Already completed in #116

Commit [2e75c4a](https://github.com/bkc39/rktorch/commit/2e75c4a) threads the MLP forward and factors MNIST into `conv-block` (convolution, activation, pooling) and a `Sequential` network. Build on these changes; do not repeat that work.

## Apply DRY thoughtfully

- Remove duplicated architectural knowledge, not merely similar-looking expressions. A useful block owns a coherent operation, its trainable state, and its invariants.
- Prefer names and boundaries recognizable from the model architecture. Keep the overall forward readable through `~>` and `Sequential`, while showing tensor mechanics in the component where they belong.
- Use a local function for stateless computation when a registered sublayer would only add indirection. Use a sublayer when parameter ownership, mode propagation, reuse, or an architectural boundary makes it worthwhile.
- Do not force superficially similar pre-norm/post-norm, self-/cross-attention, masking, or dropout behavior into one configurable abstraction prematurely.
- Keep explanatory tensor shapes visible in the literate examples. Fewer lines alone is not the success criterion.
- Preserve intentionally independent Python parity references. Sharing an implementation between the model and its reference would weaken the test; synchronized architecture does not require deduplicating across languages.
- Start with components local to their examples. Promote shared library layers only when their semantics and multiple use cases justify a stable API. Constructor-syntax changes belong in #117.

## CharGPT as the next concrete case

Audit `examples/racket/06-gpt.rkt` for these candidate boundaries:

- `CausalSelfAttention`: Q/K/V projections, score scaling, causal masking, softmax, head joining, output projection.
- `FeedForward`: expansion, GELU, and output projection.
- `PreNormResidual`: `x + branch(norm(x))`, reused around attention and the MLP.
- Consider `HeadProjection` for the repeated projection/head-splitting operation, but compare it with a local `split-heads` helper before adding another registered level.

A possible shape using the current syntax is:

```racket
(define-layer PreNormResidual (norm branch)
  #:init (width branch)
  (set! norm (LayerNorm width))
  #:forward (x)
  (~> x norm branch (+ x)))

(define-layer GPTBlock (attention mlp)
  #:init (width heads)
  (set! attention
        (PreNormResidual width (CausalSelfAttention width heads)))
  (set! mlp
        (PreNormResidual width (FeedForward width)))
  #:forward (x)
  (~> x attention mlp))
```

This is a candidate design, not a required decomposition. Review whether the named blocks improve the example as a teaching artifact. Constructor contracts must accompany production exports.

## Remaining example audit

Inspect the other literate examples, especially `07-asr.rkt`, for repeated architectural units and repeated invariants. Record what should become a local helper, an example-local layer, a shared component, or deliberate duplication. Extract only components supported by the actual model semantics; do not assume every attention implementation is interchangeable.

## Compatibility and validation

Factoring trainable children changes registration paths (for example `wq.weight` could become `attention.branch.query.projection.weight`). Decide on checkpoint compatibility explicitly: preserve paths where practical, provide a mapping/migration, or document a deliberate break. Update parameter-name assertions and the corresponding Python twins consistently.

Preserve parameter identity and sharing, traversal and seeded initialization order, buffer registration and dtypes, train/eval propagation, device behavior, and the mathematical model. Avoid mixing this structural refactor with new dropout, mask caching, or attention algorithms.

An exploratory CPU prototype of the CharGPT decomposition matched the current block's seeded parameters, outputs, and parameter gradients exactly, and Adam updates agreed. This is feasibility evidence; the final implementation still needs its own regression checks.

## Acceptance criteria

- [ ] Audit the remaining examples and document chosen boundaries, including intentional duplication.
- [ ] Refactor CharGPT into agreed architectural blocks using the merged `define-layer` API.
- [ ] Apply other justified extractions identified by the audit, without reopening completed MLP/MNIST work.
- [ ] Update literate prose and architecture diagrams alongside the code.
- [ ] Document checkpoint-name compatibility and update relevant reference implementations and assertions.
- [ ] Verify seeded initialization, forward outputs, gradients, optimizer updates, causal behavior, buffers, mode propagation, and supported-device behavior with the appropriate existing test suites and targeted regression checks.

Related: #118 proposes the optional `rktorch-resyntax` package. Establish these conventions manually first; automated threading is distinct from extracting stateful sublayers and migrating their parameter names. This issue does not block merging #116.
